lib/async.js

```diff
@@ -278,35 +281,104 @@ module.exports = function resolve(x, options, callback) {
         });
     }
 
-    function processDirs(cb, dirs) {
+    function loadManifestInDir(dir, cb) {
+        maybeRealpath(realpath, dir, opts, function (err, pkgdir) {
+            if (err) return cb(null);
+
+            var pkgfile = path.join(pkgdir, 'package.json');
+            isFile(pkgfile, function (err, ex) {
+                // on err, ex is false
+                if (!ex) return cb(null);
+
+                readFile(pkgfile, function (err, body) {
+                    if (err) cb(err);
+                    try { var pkg = JSON.parse(body); } catch (jsonErr) {}
+
+                    if (pkg && opts.packageFilter) {
+                        pkg = opts.packageFilter(pkg, pkgfile, dir);
+                    }
+                    cb(pkg);
+                });
+            });
+        });
+    }
+
+    function processDirs(cb, dirs, subpath, endsWithSubpath) {
         if (dirs.length === 0) return cb(null, undefined);
         var dir = dirs[0];
 
-        isDirectory(path.dirname(dir), isdir);
-
-        function isdir(err, isdir) {
-            if (err) return cb(err);
-            if (!isdir) return processDirs(cb, dirs.slice(1));
-            loadAsFile(dir, opts.package, onfile);
+        if (env.length > 0 && endsWithSubpath(dir)) {
+            var pkgDir = dir.substring(0, dir.length - subpath.length);
```

Comment:
```suggestion
            var pkgDir = dir.slice(0, dir.length - subpath.length);
```
(altho this one might not be equivalent; i'd prefer to avoid substring and stick to slice)